### PR TITLE
Check: re-designed

### PR DESCRIPTION
**Pull request by TikhomirovSergey**
Cases:

### - a single check without description

```java
import static org.hamcrest.Matchers.*;
import static ru.tinkoff.qa.neptune.check.CheckActionSupplier.check;
import static ru.tinkoff.qa.neptune.check.MatchAction.match;

...

var object = ...
check(object, match(is(...)));
```

Below the step hierarchy:

`Verify inspected value <here object toString value>`
|........`Check object. Assert: is <...> `

### - multiple checks without description

```java
import static org.hamcrest.Matchers.*;
import static ru.tinkoff.qa.neptune.check.CheckActionSupplier.check;
import static ru.tinkoff.qa.neptune.check.MatchAction.match;

...

var object = ...
check(object, match(is(...)),
        match(not(is(...))));
```

Below the step hierarchy:

`Verify inspected value <here object toString value>`
|........`Check object. Assert: is <...> `
|........`Check object. Assert: not is <...>`

### - multiple checks with description

```java
import static org.hamcrest.Matchers.*;
import static ru.tinkoff.qa.neptune.check.CheckActionSupplier.check;
import static ru.tinkoff.qa.neptune.check.MatchAction.match;

...

var object = ...
check("Given value", object, match(is(...)),
        match(not(is(...))));
```

Below the step hierarchy:

`Verify Given value`
|........`Check object. Assert: is <...> `
|........`Check object. Assert: not is <...>`

### - checks with evaluation

```java
import static org.hamcrest.Matchers.*;
import static ru.tinkoff.qa.neptune.check.CheckActionSupplier.check;
import static ru.tinkoff.qa.neptune.check.MatchAction.match;

...

var object = ...
check("Given value", object, match(is(...)),
        match("Some evaluated value", o -> {...}, is(...)));
```

Below the step hierarchy:

`Verify Given value`
|........`Check object. Assert: is <...> `
|........`Check Some evaluated value. Assert: is <...>`
|................`Get Some evaluated value `

###Proposed:
- re-designed code

### CODE REVIEW

File: check/src/main/java/ru/tinkoff/qa/neptune/check/MatchAction.java [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+     * @param matcher is a criteria matcher
+     * @param <T> is a type of a value to be checked
+     * @return a new {@link MatchAction}
+     */
+    public static <T> MatchAction<T, T> match(Matcher<? super T> matcher) {
+        return new DirectMatchAction<>(matcher);
+    }
+
+    /**
+     * Creates an instance that performs the matching of a value evaluated from an inspected object.
```
Comment (ArisAgnew) [inline check/src/main/java/ru/tinkoff/qa/neptune/check/MatchAction.java]: Creates an instance that performs a value matching evaluated out of inspected object.

File: check/src/main/java/ru/tinkoff/qa/neptune/check/MatchAction.java [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+
+    MatchAction(String description, Matcher<? super R> criteria) {
+        checkArgument(isNotBlank(description), "Description of a value to be evaluated and checked should not be blank");
+        checkArgument(nonNull(criteria), "Criteria matcher should not be null");
+        this.description = description;
+        this.criteria = criteria;
+    }
+
+    /**
+     * Creates an instance that performs the matching of an object.
```
Comment (ArisAgnew) [inline check/src/main/java/ru/tinkoff/qa/neptune/check/MatchAction.java]: Creates an instance that performs an object matching.

File: check/src/main/java/ru/tinkoff/qa/neptune/check/MatchAction.java [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+ * This class is designed to perform the matching of values.
+ *
+ * @param <T> is a type of an object to be checked. Also this may be some object that is used to evaluate a
+ *           value to be checked
+ * @param <R> is a generic type of {@link Matcher}
+ */
+public abstract class MatchAction<T, R> implements Consumer<T>, Supplier<Consumer<T>> {
+
+    final Matcher<? super R> criteria;
+    final String description;
```
Comment (ArisAgnew) [inline check/src/main/java/ru/tinkoff/qa/neptune/check/MatchAction.java]: private

File: check/src/main/java/ru/tinkoff/qa/neptune/check/MatchAction.java [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+/**
+ * This class is designed to perform the matching of values.
+ *
+ * @param <T> is a type of an object to be checked. Also this may be some object that is used to evaluate a
+ *           value to be checked
+ * @param <R> is a generic type of {@link Matcher}
+ */
+public abstract class MatchAction<T, R> implements Consumer<T>, Supplier<Consumer<T>> {
+
+    final Matcher<? super R> criteria;
```
Comment (ArisAgnew) [inline check/src/main/java/ru/tinkoff/qa/neptune/check/MatchAction.java]: private

File: check/src/main/java/ru/tinkoff/qa/neptune/check/InnerMatcher.java [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+import static java.lang.String.format;
+import static java.lang.System.lineSeparator;
+
+class InnerMatcher<T> extends BaseMatcher<T> {
+
+    private static final String LINE_SEPARATOR = lineSeparator();
+    private final String description;
+    private final Matcher<? super T> criteria;
+
+    InnerMatcher(String description, Matcher<? super T> criteria) {
```
Comment (ArisAgnew) [inline check/src/main/java/ru/tinkoff/qa/neptune/check/InnerMatcher.java]: Perhaps you forget to write checkArgument() methods for parameter, cuz it's an entry point of them.

File: check/src/main/java/ru/tinkoff/qa/neptune/check/EvaluateAndMatchAction.java [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+import static org.hamcrest.MatcherAssert.assertThat;
+import static ru.tinkoff.qa.neptune.core.api.steps.StoryWriter.toGet;
+
+class EvaluateAndMatchAction<T, R> extends MatchAction<T, R> {
+
+    private final Function<T, R> eval;
+
+    EvaluateAndMatchAction(Matcher<? super R> criteria, String description, Function<T, R> eval) {
+        super(description, criteria);
+        checkArgument(nonNull(eval), "Function that evaluates value to be checked should be defined");
```
Comment (ArisAgnew) [inline check/src/main/java/ru/tinkoff/qa/neptune/check/EvaluateAndMatchAction.java]: Function to evaluate a value should be defined.

File: check/src/main/java/ru/tinkoff/qa/neptune/check/EvaluateAndMatchAction.java [hunk 7]
```diff
@@ -0,0 +1,29 @@
...
+import static com.google.common.base.Preconditions.checkArgument;
+import static java.util.Objects.nonNull;
+import static org.hamcrest.MatcherAssert.assertThat;
+import static ru.tinkoff.qa.neptune.core.api.steps.StoryWriter.toGet;
+
+class EvaluateAndMatchAction<T, R> extends MatchAction<T, R> {
+
+    private final Function<T, R> eval;
+
+    EvaluateAndMatchAction(Matcher<? super R> criteria, String description, Function<T, R> eval) {
```
Comment (ArisAgnew) [inline check/src/main/java/ru/tinkoff/qa/neptune/check/EvaluateAndMatchAction.java]: It does not need at all, but in order to keep parameter position, swap description and criteria.

File: check/src/main/java/ru/tinkoff/qa/neptune/check/CheckActionSupplier.java [hunk 8]
```diff
@@ -0,0 +1,93 @@
...
+     * Creates an instance of {@link CheckActionSupplier} and defines value to be verified and verification criteria.
+     * And then the checking is performed.
+     *
+     * @param t   value to be verified.
+     * @param matchActions is an arrya of {@link MatchAction}
+     * @param <T> is a type to be verified.
+     */
+    @SafeVarargs
+    public static <T> void check(T t, MatchAction<T, ?>...matchActions) {
+        new CheckActionSupplier<T>(format("inspected value %s", t))
```
Comment (ArisAgnew) [inline check/src/main/java/ru/tinkoff/qa/neptune/check/CheckActionSupplier.java]: Inspected value
(Upper case) ?

File: check/src/main/java/ru/tinkoff/qa/neptune/check/CheckActionSupplier.java [hunk 9]
```diff
@@ -0,0 +1,93 @@
...
+                .performOn(t)
+                .get().accept(t);
+    }
+
+    /**
+     * Creates an instance of {@link CheckActionSupplier} and defines value to be verified and verification criteria.
+     * And then the checking is performed.
+     *
+     * @param t   value to be verified.
+     * @param matchActions is an arrya of {@link MatchAction}
```
Comment (ArisAgnew) [inline check/src/main/java/ru/tinkoff/qa/neptune/check/CheckActionSupplier.java]: matchActions is an array of {@link MatchAction}

File: check/src/main/java/ru/tinkoff/qa/neptune/check/CheckActionSupplier.java [hunk 10]
```diff
@@ -0,0 +1,93 @@
...
+    public static <T> void check(String description, T t, MatchAction<T, ?>...matchActions) {
+        checkArgument(!isBlank(description), "Description of a value to be inspected should not be null");
+        new CheckActionSupplier<T>(description)
+                .matches(matchActions)
+                .performOn(t)
+                .get().accept(t);
+    }
+
+    /**
+     * Creates an instance of {@link CheckActionSupplier} and defines value to be verified and verification criteria.
```
Comment (ArisAgnew) [inline check/src/main/java/ru/tinkoff/qa/neptune/check/CheckActionSupplier.java]: Creates an instance of {@link CheckActionSupplier};
Defines a value to be verified and verifivation criteria;
Value check is performed.

File: check/src/main/java/ru/tinkoff/qa/neptune/check/CheckActionSupplier.java [hunk 11]
```diff
@@ -0,0 +1,93 @@
...
+     * And then the checking is performed.
+     *
+     * @param description of a value to be verified.
+     * @param t           value to be verified.
+     * @param matchActions is an arrya of {@link MatchAction}
+     * @param <T>         is a type of a value to be verified.
+     */
+    @SafeVarargs
+    public static <T> void check(String description, T t, MatchAction<T, ?>...matchActions) {
+        checkArgument(!isBlank(description), "Description of a value to be inspected should not be null");
```
Comment (ArisAgnew) [inline check/src/main/java/ru/tinkoff/qa/neptune/check/CheckActionSupplier.java]: Value description to be inspected should not be null.

File: check/src/main/java/ru/tinkoff/qa/neptune/check/CheckActionSupplier.java [hunk 12]
```diff
@@ -0,0 +1,93 @@
...
+
+    private CheckActionSupplier(String description) {
+        super(format("Verify %s", description));
+    }
+
+    /**
+     * Creates an instance of {@link CheckActionSupplier} and defines value to be verified and verification criteria.
+     * And then the checking is performed.
+     *
+     * @param description of a value to be verified.
```
Comment (ArisAgnew) [inline check/src/main/java/ru/tinkoff/qa/neptune/check/CheckActionSupplier.java]: Value description to be verified

File: check/src/main/java/ru/tinkoff/qa/neptune/check/CheckActionSupplier.java [hunk 13]
```diff
@@ -0,0 +1,93 @@
...
+     * @param t           value to be verified.
+     * @param matchActions is an arrya of {@link MatchAction}
```
Comment (ArisAgnew) [inline check/src/main/java/ru/tinkoff/qa/neptune/check/CheckActionSupplier.java]: matchActions is an array of {@link MatchAction}

File: check/src/main/java/ru/tinkoff/qa/neptune/check/CheckActionSupplier.java [hunk 14]
```diff
@@ -0,0 +1,93 @@
...
+
+    private final List<AssertionError> caughtMismatches = new ArrayList<>();
+    private final List<Consumer<T>> checkList = new ArrayList<>();
+
+    private CheckActionSupplier(String description) {
+        super(format("Verify %s", description));
+    }
+
+    /**
+     * Creates an instance of {@link CheckActionSupplier} and defines value to be verified and verification criteria.
```
Comment (ArisAgnew) [inline check/src/main/java/ru/tinkoff/qa/neptune/check/CheckActionSupplier.java]: Creates an instance of {@link CheckActionSupplier};
Defines a value to be verified and verifivation criteria;
Value check is performed.

**Comment by TikhomirovSergey** [comment]
@ArisAgnew Take a look one more time

**Pull request merged**